### ISSUE TITLE
Need a callback when a state machine's replica state has changed

### ISSUE THREAD

**Issue by arnonmoscona** [atomix/atomix#23]
Given a replicated state machine, for instance Copycat collections, cluster members (also passive ones) don't know when the state has changed. In practical use, you create these state machines so that some kind of coordination can exist in the cluster. Cluster nodes should not need to poll the state, but rather be notified via a callback when the state has changed. So conceptually, I should be able to write something like:

```Java
AsyncSet set;
// create, open, do what you need
set.onStateChange(() -> {logger.info("State changed, now I can do something about it..."});
```

The exact mechanics are not important, maybe it's configured via `StateMachineConfig` or via some other mechanism.

**Comment by kuujo** [comment]
Great idea, and actually this is being added to Copycat right now.

I'll be finishing up the general framework for the first release of Copycat this weekend. Many changes have gone in to it, the most important of which is sessions. This was a prerequisite for things like locks and managing remote state-based notifications gracefully. Copycat will now associate state changes with a specific node and allow nodes to register state change handlers. Also, relevant is that all state changes are associated with wall clock time, so events based on expiring keys can be supported as well.

It looks like for the few weeks I'm getting some help with the release d Copycat. Right now, an initial beta release is planned for June 21st, and it will definitely have support for state change and time based events. I'll be merging those changes into master after this weekend. I'd definitely be interested to get some feedback on their implementation once that happens.

Thanks!

**Comment by arnonmoscona** [comment]
Excellent! 

Which branch is this going to go into?

BTW, I was going to do the very small bit of work needed to expose the collections' contents in the async API (e.g. `toArray()`, `stream()`) The implementation is trivial, the unit test is where a bit of work would need to get into. Should I do this work or is is already being done for the update you plan?

**Comment by kuujo** [comment]
Check out the sessions branch. That branch is the latest development branch and is quite a significant change, but it's what will ultimately  be released in a few weeks (on June 21st-ish).

The sessions branch and the first release are focused on several significant changes which I'll outline in the current README file soon (today?). Briefly, it's a significant simplification of the API (no more event log and state log, only Raft and StateMachine), major improvements to I/O (serialization framework, more efficient log and compaction), and the addition of critical missing pieces of the Raft implementation (in particular sessions).

I actually just got this branch back in to partial working state last night and tested resources, session-based state change listeners, expiring and ephemeral map keys, and other interesting features that were previously impossible, so I was just thinking about this thread when I saw this email.

It would be awesome if you could submit a PR for the stream stuff! While much of the underlying framework has changed, the AsyncMap interface itself hasn't, so you can just port over your changes. It's fine for the sessions branch right now if it doesn't have tests. I am in the process of writing tests anyways, and will be writing tests every day for the next several weeks. But AsyncMap is the only resource that does currently have a few tests in that branch, so you can add a test there if you're up to it.

**Comment by arnonmoscona** [comment]
Cool. I'll do it as soon as I can. Probably one of the evenings later this week.

**Comment by arnonmoscona** [comment]
I switched over to the sessions branch. I was working off of the master branch for my experimentation before. Looking at the sessions branch it looks like you've done a lot of work. Things changed quite a bit. I'll have to digest the new stuff. It will take a little time (doing it only on free evenings). I have a couple of questions, if you don't mind:
* Do you have any draft docs explaining the new state machine? The old simple replicated data structures are gone, and the new ones all appear to be focused on supporting ephemeral nodes, conceptually similar to Zookeeper. For my current needs I will want to recreate the simple structures using the new state machines, but the new state machines seem quite different than what I saw in master.
* Looking at `NettyClusterTest` and `ManagedMember` it looks like one must provide a numeric ID to create a member. This was not the case in the master branch. One used to provide a String that was basically a listener port spec like `"tcp:://myhost:9001"`. This form constructs a "unique ID" based on facts and limits you know at member construction time (cannot have two processes listening on this port). However, the integer ID is a "magic number". How can I know that this ID was not used elsewhere? Requiring to know or configure upfront all the possible cluster members (looks superficially looks for passive ones as well) is challenging... Any comments on this?

Please tell me if I am taking too much of your time. I would like to put some time into it, but I understand if you say that I should wait until you get the branch to a more stable state...

**Comment by kuujo** [comment]
Wow, sorry it took me so long to get to this. It seems I somehow missed this in my email. 

I'm currently working on documentation for the new state machine. It will be updated frequently until the first release.

**Question 1**
The biggest change to the new state machine is a function of the underlying log compaction algorithm. Previously, Copycat used snapshots to periodically persist state machine state and truncate the Raft log. However, there are some challenges with this approach for in-memory state machines.
* If the snapshot takes a long period of time, it can block the node. Snapshots must be associated with a specific index in the Raft log, so no new entries can be committed to the log during the snapshot period
* The alternative to pausing the log for snapshots is to partition snapshots and stream them to disk. However, this requires associating snapshot segments with specific state within the state machine, and that turns out to be very difficult to abstract for generic state machines.
* Considering that snapshots can block the Raft log, and segmenting snapshots was difficult, Copycat would have to implement leader transfer in order to force leaders to step down prior to taking a snapshot. This is a largely theoretical area of the Raft algorithm which AFAIK doesn't have any proven mechanism, only suggested ones.
* Finally, the infrequency of snapshots make for unpredictable performance.

So, rather than taking snapshots of Copycat's log, Copycat now exposes a mechanism to allow state machines to provide feedback to the log compaction process on which commands are no longer relevant to the state machine state. For instance, in the `AsyncMap` resource, given the following pair of commands:
1. `put("foo", "bar")`
2. `put("foo", "baz")`
...the state machine only needs Copycat to hold on to the second command in the event of a failure. So, the `AsyncMap` state machine implements a filter for the `Put` command indicating that entries with an index less than `2` can be safely removed from the log. Copycat will periodically compact segments of the log, prioritizing the newest segments first. When that process occurs, it allows the state machines for each command to decide whether a command should be retained in the log. 

But there are some complex challenges with this approach as well, namely tombstones. If we extend the two command example from `AsyncMap` to add a `remove("foo")` command at index `3` that poses a challenge. The map state machine no longer holds any state for the `foo` key, but the `remove("foo")` command needs to be retained in the log at least until it can be applied to all state machines in the cluster. If it's removed from one node too soon, then it may never be replicated to and applied on another node.

In order to solve that problem, Copycat separates the compaction process into two stages - `MAJOR` and `MINOR` compaction - and keeps track of which entries have been replicated to all nodes in the cluster. During major compaction, Copycat compacts the entire log from the head (index `1`) to the index of the most recent entry that has been replicated to all nodes in the cluster. During the major compaction process, tombstones can be safely dropped from the log. The reason that the compaction has to take place on the entire log from head to tail is because it needs to ensure that all entries that contributed to the state prior to the tombstone are removed *before* the tombstone can be removed, otherwise state machines could diverge from one another depending on the compaction process, and the key on one node could persist while the key on another node could have been compacted out of the log.

This all sounds like more complexity than snapshotting, but it makes for a much smoother system with more predictable performance. Log compaction is made extremely efficient by performing only sequential reads and writes and writing large segments to disk at a time. The compaction process can be throttled to reduce the load on the server without impacting the operation of the Raft algorithm itself.

Additionally, higher level abstractions can be built on top of the log cleaning algorithm. For instance, Kafka's log cleaning algorithm uses entry keys to deduplicate entries in the log, and `null` values to indicate tombstones. A similar abstraction can be built on top of Copycat's `StateMachine` in a `KeyedStateMachine` that overrides the `filter` method to filter out commands by key. So, while Copycat's core state machines like `AsyncMap` and others will always extend the low level `StateMachine` class for more control, users should be given tools that can help more easily work within Copycat's log compaction model.

**Question 2**

The topic of member configurations is a complex one. The challenge of the Raft algorithm is that Copycat needs to have explicit information about all Raft voting members that is consistent across all nodes. Previously, Copycat did that using URIs, which was't such a bad model. But the reasoning behind using `int` IDs instead of URIs is for portability.

Currently, Copycat does not support cluster membership changes. This poses a particular challenge in the event that a node permanently dies. Imagine a scenario where I start three nodes, one a `123.456.789.0`, one at `123.456.789.1`, and one at `123.456.789.2`. If the first node is hit by an asteroid, it will be very difficult to recover it as a member of the Raft cluster. What `int` IDs allow is for nodes to arbitrarily change hosts or ports without altering the cluster membership. So, even in a scenario where I simply want to change the port on which a Copycat server is listening, given that the server has a member ID I can simply restart it on a new port, and the leader will associate state from the old member with state from the new member.

Coincidentally, ZooKeeper does the same thing with node IDs and probably for similar reasons. ZooKeeper requires that nodes participating in the consensus algorithm have a `myid` file which indicates the node's numeric ID. I suspect a similar pattern could be used for configuring Copycat nodes, but I've opted to leave that element to users and/or a future runnable server.

That said, I think I will definitely revisit the concept of node IDs in Copycat. Perhaps the server address can still be used the the caveat that changes to the host or port will result in changes to the Copycat cluster. I expect that we'll be implementing full support for membership changes fairly soon. See the below update for details.

**On Copycat's status**
After running and passing some basic [Jepsen](https://github.com/aphyr/jepsen) tests to verify linearizability over the past couple of weeks, I just merged the `0.6.0` branch into `master` and will be continuing to work towards a release. This version has been shown to be significantly more stable than any previous development version, but there's still a lot of work remaining to do. We've been able to uncover a new issue in Jepsen testing that I'm working to resolve right now. Additionally, there is a lot of profiling and performance tuning to be done for log compaction and reference counting processes, and of course there are a large number of tests that remain to be written. A full release will not happen until significant tests are written, all Jepsen tests pass, all interfaces are documented, and the system is more thoroughly profiled, but we have made significant strides in verifying the algorithm, protocol, and log in a distributed environment, and the more people that play with it now, the more quickly it will become stable.

Don't worry about taking too much of my time! I love talking about this stuff, and I'm big on documentation, so I'm happy to explain things in the meantime. Honestly, these types of conversations help me sort out what needs to be explained in the documentation and how to prioritize new features (like a simpler `StateMachine` base class). Plus I might be able to steal some of my explanation for the documentation.
